### PR TITLE
Scoped values

### PR DESCRIPTION
**Pull request by lortabac**
This proposal introduces `ScopedValues`, a new GHC extension
that provides first-class support for dynamic scoping with robust and predictable semantics.

[Rendered](https://github.com/lortabac/ghc-proposals/blob/nominal-implicit-parameters/proposals/0741-scoped-values.rst)

### CODE REVIEW

**Comment by lortabac** [comment]
@arybczak 

> it's a far better solution to fix quirks of the existing implementation than trying to duplicate it as a separate extension with slightly different semantics.

I don't disagree, but fixing a feature that has been there for >10 years without causing damage is beyond my ability. I prefer sticking to what I can realistically achieve.

File: proposals/0741-scoped-values.rst [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+
+Now unrestricted IO is possible, but it must happen within a ``runIO`` continuation
+and requires the ``HasIO`` constraint:
+
+::
+
+  launchMissiles :: HasIO => Eff ()
+  launchMissiles = performIO $ putStrLn "Missiles launched!"
+
+  >>> runEff $ runIO $ runTime defaultTime (launchMissiles >> getCurrentTime)
```
Comment (marinelli) [inline proposals/0741-scoped-values.rst]: `currentTime`

**Comment by blamario** [comment]
I don't think that dynamic scoping should be marketed as beginner-friendly in any form. That aside, I have a couple of syntactic quibbles:
* `scoped` is misleading, in that all user-defined names in Haskell are already scoped. The distinction here is that the scope is dynamic
* `data ... =` is generally followed by a constructor, not a type. You'd stay closer to Haskell syntax if you replaced `data scoped` with `type scoped`
* Combining the two preceding objections, how about `type dynamic K = Int` or even `type instance Dynamic K = Int`? Note the latter doesn't even require a syntactic extension.

**Comment by lortabac** [comment]
I apologize, I realized that the proposal is completely silent on this point. I tried to avoid talking about implementation but now it seems unavoidable.
Also, I should point out that of course I'm not an expert GHC contributor. This idea is based on what I understood from my experiments with the GHC code base, but I may be missing some important caveats.

The trick is that `Scoped` constraints are resolved eagerly. The solver resolves `Scoped` wanteds by picking the innermost given at the point where the wanted arises.
Consider this example:

```
test :: String
test = withValue @Scope "global" $
  let capture = value @Scope
  in withValue @Scope "local" capture
```

When typechecking `let capture = value @Scope`, the solver sees a wanted `Scoped Scope` and immediately resolves it against the given from `withValue @Scope "global"`. At that point `capture` gets a concrete `String` type. During generalization there's no `Scoped` constraint left to generalize over, because it was already solved. So, whether MR is enabled or not, `capture` always has type `String`.

This has an important downside: scoped values always require a signature. The following code will not compile, because the `Scoped P` constraint must appear in the signature of `getP`.

```
data scoped P = Int

getP = value @P
```

I think I should mention all this in the proposal.

**Comment by tomjaguarpaw** [comment]
>  You can specify what happens (regardles of flags) simply by giving a type signature.

Let's look and see what happens if we try this. Below we see that we get `"global"` or `"local"` according to whether `(?scope :: String) =>` appears in the type signature or not.

```.hs
{-# LANGUAGE ImplicitParams #-}

-- ghci> foo
-- "global"
foo :: String
foo = let ?scope = "global" in
  let capture :: String
      capture = ?scope
  in let ?scope = "local" in capture
```

```.hs
-- ghci> foo
-- "local"
foo :: String
foo = let ?scope = "global" in
  let capture :: (?scope :: String) => String
      capture = ?scope
  in let ?scope = "local" in capture
```

Alright, so we can disambiguate this usage of `ImplicitParams` using an explicit type signature.

----

How about if we try the same thing using @lortabac's proposal here? Would both of these be valid programs? How would their results differ?

```.hs
test :: String
test = withValue @Scope "global" $
  let capture :: String
      capture = value @Scope
  in withValue @Scope "local" capture
```

```.hs
test :: String
test = withValue @Scope "global" $
  let capture :: Scoped Scope => String
      capture = value @Scope
  in withValue @Scope "local" capture
```

Which one do we get (or do we get something else?) if the signature is left off? (Above you say "scoped values always require a signature" but I think you mean at the top level, right? At least the example `test` above that has `let capture = value @Scope` without a signature.)

**Comment by lortabac** [comment]
@tomjaguarpaw If there is a signature it behaves the same as `ImplicitParams`. See https://github.com/lortabac/ghc-proposals/blob/nominal-implicit-parameters/proposals/0741-scoped-values.rst#local-capture

The only difference is that if the local binding has no signature the value is always captured locally, independently of the enabled extensions.

> Above you say "scoped values always require a signature" but I think you mean at the top level, right?

Yes. I meant at top level. If there is a `Scoped` wanted there must be a matching given, and my understanding is that at top-level it can only be provided by a signature.

**Comment by lortabac** [comment]
> my understanding is that at top-level it can only be provided by a signature

Unless we allow global/default values, of course. But I excluded that feature for simplicity and predictability.

**Comment by tomjaguarpaw** [comment]
> If there is a signature it behaves the same as `ImplicitParams` ... The only difference is that if the local binding has no signature the value is always captured locally

Ah, OK, so it looks like the situation is as follows. Is that right?

### `ImplicitParams`

```.hs
  let capture :: String
      capture = ?scope
```

:point_up_2: captures

```.hs
  let capture :: (?scope :: String) => String
      capture = ?scope
```

:point_up_2: doesn't capture

```.hs
  let capture = ?scope
```

:point_up_2: depends on other extensions

### `ScopedValues`

```.hs
  let capture :: String
      capture = value @Scope
```

:point_up_2: captures

```.hs
  let capture :: Scoped Scope => String
      capture = value @Scope
```

:point_up_2: doesn't capture

```.hs
  let capture = value @Scope
```

:point_up_2: captures

**Comment by lortabac** [comment]
@tomjaguarpaw Yes!

There is also an edge case, which affects both `ImplicitParams` and `ScopedValues`.
When you provide a type annotation instead of a full signature, it behaves as if no signature was provided:

```
test :: String
test = let ?scope = "global" in
  let (capture :: (?scope :: String) => String) = ?scope
  in let ?scope = "local" in capture
```

Depends on the other extensions :arrow_up: 

```
test :: String
test = withValue @Scope "global" $
  let (capture :: Scoped Scope => String) = value @Scope
  in withValue @Scope "local" capture
```

Always captures :arrow_up:

**Comment by lortabac** [comment]
Interestingly, `MonoLocalBinds` doesn't seem to have any effect on the behavior of `ImplicitParams` (I only tested with GHC 9.12.2).
Which is reassuring and scary at the same time, because on the one hand it means that toggling eg. `TypeFamilies` doesn't break code; on the other hand it is surprising and lets you wonder if this behavior is reliable and "guaranteed" across GHC versions.

**Comment by simonpj** [comment]
> Interestingly, MonoLocalBinds doesn't seem to have any effect on the behavior of ImplicitParams (I only tested with GHC 9.12.2).

The the same rules apply as to any other local binding

* With `MonoLocalBinds` a local definition is generalised unless it has not locally-bound free term variables.  `capture` does not (`?scope` is not treated as a locally bound free variable), so it is generalised.  I was wrong to point to `MonoLocalBinds`.

* With `MonomorphismRestriction` (on by default) a local binding is not generalised over any constraints.  So `capture` is not generalised over `?scope =>`, and has type `capture :: String`

So if you compile
```
foo :: String
foo = let ?scope = "global" in
  let capture  = ?scope
  in let ?scope = "local"
  in capture 
```
* With `MonomorphismRestriction` you get "global"
* With `NoMonomorphismRestriction` you get "local"

These rules are not simple, but they are predictable, reliable, and I don't believe they have changed for a couple of decades.

GHC has Note about this behaviour: `Note [Inheriting implicit parameters]`.

-----------------

One might question whether GHC's rules for implicit parameters are the "right" ones, although in 30 yrs I have never had a bug report that turned out to be a misundersatnding about implcit parameter capture.  But even if there was a sold consensus about some alternative design, I'm cautious about desiging a whole new, 90% overlapping, feature in response.

**Comment by lortabac** [comment]
Let's recapitulate what this proposal is trying to fix. I think we can classify the issues in two categories: blocking points, which make implicit parameters unusable for some tasks, and "nice to have"s.

The blocking points, due to the fact that parameters are strings:

- Naming conflicts
- Lack of abstraction

The minor issues:

- Unused and shadowed parameter warnings
- The ability to silence "shadowed parameter" warnings for a specific parameter.
- Forbidding global instances
- Making local capture robust against extensions

Which of these things can be retrofitted to the existing implicit parameters implementation? How much breakage would the change cause?
Maybe trying to answer these questions can be helpful in determining whether a new extension is really worth it or not.

**Comment by adamgundry** [comment]
I'm not convinced about the language-level changes here (though I think library/compiler changes may be more reasonable). In particular, the `data scoped` stuff seems a bit too much complexity to add IMHO given the motivation. There are other motivations for allowing user-defined open data kinds, rather having to use encodings that target one of the existing open data kinds, but that would need something more general.

What if we were to introduce a new built-in class similar to `IP`, like this?
```hs
type Scoped :: Type -> Constraint -- Perhaps representation-polymorphic?
class Scoped t where
  value :: t

withValue :: t -> (Scoped t => r) -> r
withValue = GHC.Exts.withDict
```
GHC could prohibit user-defined instances of the `Scoped` class, implement special treatment in the constraint solver much as it does for `IP`, and issue shadowing/unused-constraint warnings as needed.

Then:
 * `data scoped K = Int` becomes `newtype K = K Int`
 * `withValue @K 1` becomes `withValue (K 1)`
 * `value @K` is unchanged

This seems rather simpler, but still covers most of the desiderata, I think? It does have the downside that users are not forced to introduce newtypes by the interface, and may instead work directly with `Scoped Int`.

**Comment by lortabac** [comment]
@adamgundry Is it worth introducing a new type class though? The "special treatment" for `IP` would have to be duplicated for `Scoped`.
Once you give up on solving the monomorphism restriction problem, then the only difference between `IP` and `Scoped` is in the signature. If it's possible to make the first argument of `IP` poly-kinded (as suggested in the discussion), then I think there is no need for a new class. `withValue` and `value` can use `withDict` and `ip` and be defined at library level. The API would be slightly different, but equivalent. Something like:
```
withValue :: forall x -> a -> (IP x a => r) -> r

value :: IP x a => forall x -> a
```
The only missing feature, compared to your solution, would be disallowing `IP` instances, but I suppose that is a minor issue.
Am I missing anything? Also, is it possible to make `IP` poly-kinded or are `Symbol`s somehow fundamental in the implementation?

**Comment by adamgundry** [comment]
> Is it worth introducing a new type class though? The "special treatment" for `IP` would have to be duplicated for `Scoped`.

Fair question. I was envisaging that a new class would allow the constraint solving / generalization strategy to be different, and allow forbidding global instances, whereas making those changes for `IP` would have tricky backwards compatibility implications. But if we're keeping broadly the same the constraint solver behaviour for `IP`, then I agree a new class is less motivated. I don't have a clear view on whether the hesitancy about using IPs is mainly related to the fact they are strings, or about constraint solving issues.

Regarding global instances of `IP`, perhaps it would be reasonable to emit a warning about them (which can be disabled if the user really needs a global instance)?

> is it possible to make `IP` poly-kinded or are `Symbol`s somehow fundamental in the implementation?

I have not looked closely at the implementation, but my guess is that it is probably not vastly difficult to make `IP` poly-kinded while preserving backwards compatibility in normal cases. We did something similar to `HasField` before, and there is a similar story: when the compiler generates/solves constraints arising from IP-specific syntax it will do so with `Symbol`s, but there's nothing particularly problematic about the underlying class being poly-kinded and merely instantiating it to `Symbol` in the relevant cases. It might affect code that works directly with `IP`, but `IP` is supposed to be an internal implementation detail anyway.

---

Another thought: I have [wondered previously](https://gitlab.haskell.org/ghc/ghc/-/issues/23831) about adding a pragma or similar mechanism to explicitly mark classes that are not globally coherent but are used to pass context-dependent implicit arguments. That is, rather than `IP` being deeply magical, we could allow any class to be given similar properties. This would turn the question of `IP` vs `Scoped` into a library design issue, which should help reduce coupling between GHC and the standard libraries. And I think we could allow an IP to be a superclass of such a class? I have not thought deeply about the design here though.

**Comment by lortabac** [comment]
To be honest I think that `withDict` is currently too unsafe to be advertised as the recommended way to pass implicit context.

However `withDict`, in combination with the `NON_SINGLETON` pragma, may constitute a generalized version of this proposal if some restrictions are introduced.
- `withDict` should only be allowed on `NON_SINGLETON` classes. This is to prevent breaking global coherence and do things like (`withDict @Hashable (const 1) $ Map.lookup k m`).
- `NON_SINGLETON` classes should not be allowed to have global instances. Or, at the very least, the resolution logic should be predictable and documented in those cases.
- `NON_SINGLETON` classes should not be allowed as super-classes of ordinary type classes.
- There should be warnings to prevent confusing cases.
- Ideally the generalization strategy of `NON_SINGLETON` classes should not depend on the enabled extensions.

This is basically identical to my proposal (minus the surface syntax), but instead of relying on a magic `Scoped` constraint it generalizes this behavior to more classes. It reduces the coupling between the compiler and the libraries, as you say.

@adamgundry What do you think?

**Comment by adamgundry** [comment]
That does seem like a promising approach to me (although I have not thought through all the implications). If you're willing to do explore it further, it could definitely do with being fleshed out into a full proposal.

> * `withDict` should only be allowed on `NON_SINGLETON` classes. This is to prevent breaking global coherence and do things like (`withDict @Hashable (const 1) $ Map.lookup k m`).

This is a bit subtle, because AFAIK there are various cases where `withDict` is  "safe":
 - The compiler knows the class is not globally coherent (e.g. because it was marked as `NON_SINGLETON`) and accordingly refrains from specialization or otherwise assuming coherence.
 - The type stored in the class has at most one inhabitant (e.g. `SingI` from singletons).
 - The class plays higher-rank polymorphism tricks (like `Reifies` from `reflection`).
 - For each type at which the user program calls `withDict`, there is no meaningful difference between the values used as arguments in calls to `withDict`. (This is trivially true if there is only one call to `withDict` in the program, but it may also be the case that the user program doesn't distinguish between certain values.)

So there's a non-obvious question of where `withDict` should be silently accepted, where it should result in a warning, and where it should be rejected (albeit perhaps with an `unsafeWithDict` alternative). The last point in particular can likely only be left to "trust the user not to misuse this".

> * `NON_SINGLETON` classes should not be allowed to have global instances. Or, at the very least, the resolution logic should be predictable and documented in those cases.

I don't think global instances are inherently problematic in themselves. Assuming the `NON_SINGLETON` contract means "this class is not globally coherent", global instances are just another potential source of incoherence. But agreed the resolution logic needs to be clearly specified. (If a particular class wants to forbid global instances, that can be done by having a non-exported superclass.)

> * `NON_SINGLETON` classes should not be allowed as super-classes of ordinary type classes.

A key benefit of `reflection` over implicit parameters is that `Reifies` *can* be used as a superclass, and ideally this story would allow `reflection` to use `withDict`. But perhaps we should say that `NON_SINGLETON` classes cannot be superclasses of globally coherent classes, `Reifies` is *not* a `NON_SINGLETON` class, but `withDict` *can* be used for it?

> * There should be warnings to prevent confusing cases.

Absolutely, though we might have to pin down what this means. :grin: 

> * Ideally the generalization strategy of `NON_SINGLETON` classes should not depend on the enabled extensions.

Indeed, though here we're back into "what do we do about `IP`" territory again, because it would be nice if we end up with a story that makes `IP` consistent with everything else. But to begin with it would be good to specify what generalization strategy we really want for `NON_SINGLETON`, and how that differs from what `IP` currently does.

Side note: `NON_SINGLETON` is a terrible name I came up with without thinking, and I guess these days a Modifier would be preferred to a pragma. So feel free to change it (though I guess picking a bikeshed colour can wait until the overall story is clearer).

**Comment by lortabac** [comment]
After reflecting on the discussion, I've decided to close this proposal.

The idea has proven more controversial than I thought. I don't think it's a good idea to insist if the consensus is not broad enough.

The more general solution that came up in conversation with @adamgundry seems like it could address the problem more fundamentally and subsume this proposal. That's a significantly harder undertaking though, and I'm not ready to tackle it right now. I may try in the future. Or, if anyone wants to explore that direction, I'd be happy to contribute to the discussion.

**Comment by VitWW** [comment]
> After reflecting on the discussion, I've decided to close this proposal.

You need explicit push the button "Close pull request" to close the proposal.
